### ISSUE TITLE
[YSQL] Avoid regenerating colocation id for same table

### ISSUE THREAD

**Issue by jasonyb** [yugabyte/yugabyte-db#14239]
Jira Link: [DB-3697](https://yugabyte.atlassian.net/browse/DB-3697)
### Problem

There was a user issue where tserver check failed in `RaftGroupMetadata::AddTable`:

```cpp
      CHECK(!existing_schema.has_colocation_id() ||
            existing_schema.colocation_id() == schema.colocation_id())
              << "Attempted to change colocation ID for table " << table_id
              << " from " << existing_schema.colocation_id()
              << " to " << schema.colocation_id();
```

In debug build, an earlier check is hit:

```cpp
      LOG(DFATAL) << "Table " << table_id << " already exists. New table info: "
          << new_table_info->ToString() << ", old table info: " << existing_table.ToString();
```

This means that the metadata for that table id already existed, there was a request to ADD_TABLE for that table id, but the colocation id differs.  Colocation id is generated in catalog manager during `CreateTable`.  My guess on how this happened is that client (table_creator.cc) tried the CreateTable RPC to master twice, maybe because the first one got lost when returning back to the client.

I wasn't able to reproduce when using tricks like iptables to cause packet loss.  I'm not sure what the exact mechanism is for the retry of `SyncLeaderMasterRpc`.  I ended up manually modifying the code to simulate a retry:

```diff
diff --git i/src/yb/client/table_creator.cc w/src/yb/client/table_creator.cc
index b0e62fded2..4c4b4fc851 100644
--- i/src/yb/client/table_creator.cc
+++ w/src/yb/client/table_creator.cc
@@ -346,6 +346,15 @@ Status YBTableCreator::Create() {
       RETURN_NOT_OK_PREPEND(s, strings::Substitute("Error creating $0 $1 on the master",
                                                    object_type, table_name_.ToString()));
   }
+  if (req.is_colocated_via_database()) {
+    s = client_->data_->CreateTable(
+        client_, req, *schema_, deadline, &table_id_);
+
+    if (!s.ok() && !s.IsAlreadyPresent()) {
+        RETURN_NOT_OK_PREPEND(s, strings::Substitute("Error creating $0 $1 on the master",
+                                                     object_type, table_name_.ToString()));
+    }
+  }
 
   // We are here because the create request succeeded or we received an IsAlreadyPresent error.
   // Although the table is already in the catalog manager, it doesn't mean that the table is
```

With rf 3 cluster and create table in colocated database, immediately all tservers FATALed hitting that check (this was fastdebug build)

```
F0928 16:22:22.835664 1000838 tablet_metadata.cc:984] Table 00004000000030008000000000004001 already exists. New table info: table_id: "00004000000030008000000000004001" table_name: "t" table_type: PGSQL_TABLE_TYPE schema { columns { id: 10 name: "ybrowid" type { main: BINARY } is_key: true is_nullable: false is_static: false is_counter: false sorting_type: 1 order: -100 pg_type_oid: 20 } columns { id: 11 name: "i" type { main: INT32 } is_key: false is_nullable: true is_static: false is_counter: false sorting_type: 0 order: 1 pg_type_oid: 23 } table_properties { contain_counters: false is_transactional: true consistency_level: STRONG use_mangled_column_name: false is_ysql_catalog_table: false retain_delete_markers: false partitioning_version: 1 } colocated_table_id { colocation_id: 1863040564 } pgschema_name: "public" } schema_version: 0 partition_schema { range_schema { columns { id: 10 } } } namespace_name: "" wal_retention_secs: 0, old table info: table_id: "00004000000030008000000000004001" table_name: "t" table_type: PGSQL_TABLE_TYPE schema { columns { id: 10 name: "ybrowid" type { main: BINARY } is_key: true is_nullable: false is_static: false is_counter: false sorting_type: 1 order: -100 pg_type_oid: 20 } columns { id: 11 name: "i" type { main: INT32 } is_key: false is_nullable: true is_static: false is_counter: false sorting_type: 0 order: 1 pg_type_oid: 23 } table_properties { contain_counters: false is_transactional: true consistency_level: STRONG use_mangled_column_name: false is_ysql_catalog_table: false retain_delete_markers: false partitioning_version: 1 } colocated_table_id { colocation_id: 2152965895 } pgschema_name: "public" } schema_version: 0 partition_schema { range_schema { columns { id: 10 } } } namespace_name: "" wal_retention_secs: 0
```

Notice that colocation_id differs between the two.

### Solution

If the client generated the colocation id, this issue could have been avoided (in release build).  An issue with generating it on the client is that that colocation id may already be taken (master knows), but in that case, master responds with InvalidArgument about that and the client could just generate another colocation id.

Also, in general, it would be helpful to have colocation id generation be "mostly deterministic" by doing something like trying colocation_id = table id, and if that id's already taken, generate a random one.  That could make debugging easier in the common case.

### Other

- It would also be good to avoid sending the same ADD_TABLE from master to tservers in the first place.  This may not be possible to 100% guarantee, so the DFATAL should probably change to only FATAL when the schemas differ.
- This might be easy to reproduce with backup-restore.  I forgot the order, but if snapshot is restored that contains a table with colocation id X and DDL is run for the same table with colocation id Y, this may hit a similar error.

**Comment by jasonyb** [comment]
I'm seeing several hits of the DFATAL in jenkins debug build logs.  I checked one of them and found that the "New table info" and "old table info" completely match.  This adds to the argument that the DFATAL should be relaxed.  Or, some other thing should be implemented to prevent reaching this case in the first place, if possible.

**Comment by tverona1** [comment]
I was able to repro this scenario after hitting a create index timeout on a debug build, RF=1:
```
-- on colocated database:
CREATE TABLE test111 WITH (colocated = true) AS SELECT c_int, (case when c_int % 2 = 0 then true else false end) as c_bool, (c_int + 0.0001)::text as c_text, (c_int + 0.0002)::varchar as c_varchar, (c_int + 0.1)::decimal as c_decimal, (c_int + 0.2)::float as c_float, (c_int + 0.3)::real as c_real, (c_int + 0.4)::money as c_money FROM generate_Series(1,1200000) c_int;

create index i_114 on test111 (c_int asc);
ERROR:  Timed out: Timed out waiting for Create Table

```

After this, a subsequent create index on the same table fails with:
```
F20221207 12:33:42 ../../../../../src/yb/tablet/tablet_metadata.cc:1025] Table 00004113000030008000000000004305 already exists. New table info: table_id: "00004113000030008000000000004305" table_name: "i_114" table_type: PGSQL_TABLE_TYPE schema { columns { id: 10 name: "c_int" type { main: INT32 } is_key: true is_nullable: false is_static: false is_counter: false sorting_type: 3 order: 1 pg_type_oid: 23 } columns { id: 11 name: "ybidxbasectid" type { main: BINARY } is_key: true is_nullable: false is_static: false is_counter: false sorting_type: 1 order: -101 pg_type_oid: 0 } table_properties { contain_counters: false is_transactional: true consistency_level: STRONG use_mangled_column_name: false is_ysql_catalog_table: false retain_delete_markers: true partitioning_version: 1 } colocated_table_id { colocation_id: 1004511590 } pgschema_name: "public" } schema_version: 0 partition_schema { range_schema { columns { id: 10 } columns { id: 11 } } } namespace_name: "" wal_retention_secs: 0 namespace_id: "", old table info: table_id: "00004113000030008000000000004305" table_name: "i_114" table_type: PGSQL_TABLE_TYPE schema { columns { id: 10 name: "c_int" type { main: INT32 } is_key: true is_nullable: false is_static: false is_counter: false sorting_type: 3 order: 1 pg_type_oid: 23 } columns { id: 11 name: "ybidxbasectid" type { main: BINARY } is_key: true is_nullable: false is_static: false is_counter: false sorting_type: 1 order: -101 pg_type_oid: 0 } table_properties { contain_counters: false is_transactional: true consistency_level: STRONG use_mangled_column_name: false is_ysql_catalog_table: false retain_delete_markers: true partitioning_version: 1 } colocated_table_id { colocation_id: 1004511590 } pgschema_name: "public" } schema_version: 0 partition_schema { range_schema { columns { id: 10 } columns { id: 11 } } } namespace_name: "" wal_retention_secs: 0 namespace_id: ""
    @        0x10cfc60dc  google::LogDestination::LogToSinks()
    @        0x10cfc5238  google::LogMessage::SendToLog()
    @        0x10cfc5bdc  google::LogMessage::Flush()
    @        0x10cfc59e0  google::LogMessage::~LogMessage()
    @        0x108b90f24  yb::tablet::RaftGroupMetadata::AddTable()
    @        0x1088ff6a0  yb::tablet::Tablet::AddTableInMemory()
    @        0x1088ff9b0  yb::tablet::Tablet::AddTable()
```

**Comment by yifanguan** [comment]
Tomer's repro has an index creation timeout issue as a prerequisite, which was fixed by commit e525416774d9.

**Comment by yifanguan** [comment]
Issue resolved by commit 3df1450
